### PR TITLE
Support JDK9 methods using MR-JAR

### PR DESCRIPTION
**Pull request by FroMage**
WDYT @manovotn ? I'm not too sure how to test this though. I have to check how other projects do it in the CI.

### CODE REVIEW

File: core/pom.xml [hunk 1]
```diff
@@ -9,6 +10,71 @@
...
+
+
+    <build>
+        <plugins>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-antrun-plugin</artifactId>
+                <executions>
+                    <execution>
+                        <id>compile-java9</id>
```
Comment (manovotn) [inline core/pom.xml]: Shouldn't we now enforce min JDK version to build the project?

File: tests/src/test/java/io/smallrye/context/test/MiscTest.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+import org.eclipse.microprofile.context.ThreadContext;
+import org.eclipse.microprofile.context.spi.ContextManager;
+import org.eclipse.microprofile.context.spi.ContextManagerProvider;
+import org.junit.Assert;
+import org.junit.Test;
+
+public class MiscTest {
+
+    @Test
+    public void testCFWrapping() {
```
Comment (manovotn) [inline tests/src/test/java/io/smallrye/context/test/MiscTest.java]: What does this test exactly? Isn't the assert always true?

File: core/pom.xml (hunk 1 above)
Comment (FroMage) [reply core/pom.xml]: Ah, good point.

File: tests/src/test/java/io/smallrye/context/test/MiscTest.java (hunk 2 above)
Comment (manovotn) [reply tests/src/test/java/io/smallrye/context/test/MiscTest.java]: Not sure I follow, can you expand please?

File: core/pom.xml [hunk 3]
```diff
@@ -9,6 +10,71 @@
...
+                        <phase>compile</phase>
+                        <configuration>
+                            <tasks>
+                                <property name="compile_classpath" refid="maven.compile.classpath" />
+                                <mkdir dir="${java9.build.outputDirectory}" />
+                                <javac fork="true" executable="${env.JAVA_HOME}/bin/javac" srcdir="${java9.sourceDirectory}"
```
Comment (manovotn) [inline core/pom.xml]: Also relevant to my previous comment, assuming we will enforce building on JDK 9+ (most likely 11), then you shouldn't need this. Having to have env variable defined to make a build pass is sort of weird, ideally it should all "just work" with `mvn clean install` ;-)

**Comment by FroMage** [comment]
@manovotn I've updated the way I do the MR to the first option recommended on https://maven.apache.org/plugins/maven-compiler-plugin/multirelease.html which means we have two virtual modules:

- `core-java8` (the base)
- `core-java9` (the MR for 9)

and `core` just assembles the two into an MR jar.

It's more standard build for Maven I guess, but it has two drawbacks:

- Eclipse can't load the modules properly because it doesn't see that the `core` module has classes, so you need to add the `core-java8/9` projects to its build path.
- Dependencies from `core-java8` need to be duplicated in `core` because that's the pom that will be used.

I added tests for the new methods too.

**Comment by FroMage** [comment]
Updated it again with the structure supported by smallrye-parent, that's much cleaner.

File: core/bin/mrjar.xml [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+        <includeDependencies>false</includeDependencies>
+        <unpackOptions>
+          <excludes>
+            <exclude>/META-INF/**</exclude>
+          </excludes>
+        </unpackOptions>
+      </binaries>
+    </moduleSet>
+  </moduleSets>
+</assembly>
```
Comment (Ladicek) [inline core/bin/mrjar.xml:37]: Is this file still used?

File: core/src/main/resources/META-INF/beans.xml [hunk 5]
```diff
@@ -0,0 +1,12 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<beans xmlns="http://java.sun.com/xml/ns/javaee" 
+       xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" 
+       xmlns:weld="http://jboss.org/schema/weld/beans" 
+       xsi:schemaLocation="
+          http://java.sun.com/xml/ns/javaee http://docs.jboss.org/cdi/beans_1_0.xsd
+          http://jboss.org/schema/weld/beans http://jboss.org/schema/weld/beans_1_1.xsd">
+    <weld:scan>
+        <!-- Don't let Arquillian choke on MR-Jars -->
```
Comment (manovotn) [inline core/src/main/resources/META-INF/beans.xml:9]: Hmm this should be fixed in more recent version of the container - https://github.com/arquillian/arquillian-container-weld/commit/7296f97caac5ceaf8a5ec3a11a483a7c36e84a44#diff-797455798ac6dee103ff0fd610638bb5
Version of arq container seems to come from SR parent and newer versions are already updated, just bump it and you can remove this file :)

File: core/bin/mrjar.xml (hunk 4 above)
Comment (manovotn) [reply core/bin/mrjar.xml:37]: This should be safe to remove.

File: tests/src/test/java/io/smallrye/context/test/CompletableFutureTest.java [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+                })
+                .get();
+    }
+
+    @Test
+    public void testJava9Methods() throws InterruptedException, ExecutionException {
+        try {
+            CompletableFuture.class.getDeclaredMethod("copy");
+        } catch (NoSuchMethodException | SecurityException e) {
+            Assume.assumeTrue(false);
```
Comment (manovotn) [inline tests/src/test/java/io/smallrye/context/test/CompletableFutureTest.java:116]: I know it's nitpicking but why not just `return`?
Also, a comment explaining why this is ok might be good.

File: core/src/main/resources/META-INF/beans.xml (hunk 5 above)
Comment (manovotn) [reply core/src/main/resources/META-INF/beans.xml:9]: Actually there is already https://github.com/smallrye/smallrye-context-propagation/pull/181
So you can choose between merging and rebasing or updating it as part of this PR, up to you.

File: tests/src/test/java/io/smallrye/context/test/CompletableFutureTest.java (hunk 6 above)
Comment (FroMage) [reply tests/src/test/java/io/smallrye/context/test/CompletableFutureTest.java:116]: If you return the test is not marked as skipped.

**Comment by FroMage** [comment]
I think this PR's CI is toast, lemme reopen one.

File: tests/src/test/java/io/smallrye/context/test/CompletableFutureTest.java (hunk 6 above)
Comment (FroMage) [reply tests/src/test/java/io/smallrye/context/test/CompletableFutureTest.java:117]: I added a comment.

File: core/src/main/resources/META-INF/beans.xml (hunk 5 above)
Comment (FroMage) [reply core/src/main/resources/META-INF/beans.xml]: Well spotted, I've merged #181, rebased and removed that file.

File: core/bin/mrjar.xml (hunk 4 above)
Comment (FroMage) [reply core/bin/mrjar.xml]: Ah yes, thanks.

File: tests/src/test/java/io/smallrye/context/test/CompletableFutureTest.java (hunk 6 above)
Comment (manovotn) [reply tests/src/test/java/io/smallrye/context/test/CompletableFutureTest.java:117]: > If you return the test is not marked as skipped.

Ah, fair enough, didn't realize that.

**Comment by FroMage** [comment]
Finally, CI rules are fixed.

**Comment by dmlloyd** [comment]
> I am probably missing something but why do you need empty files such as `core/build-release-8` or `tests/build-release-8`?
> 
> Otherwise looks pretty good.

Have a look at the docs here: https://github.com/smallrye/smallrye-parent/blob/master/README.adoc

**Pull request merged**